SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityTaskShuffler.java

```diff
@@ -0,0 +1,279 @@
+        break;
+      }
+
+      long shufflingTasksOnSlave = 0;
+      double cpuUsage = getSystemLoadForShuffle(slave.usage);
+      double memUsageBytes = slave.usage.getMemoryBytesUsed();
+
+      TaskCleanupType shuffleCleanupType = slave.resource.toTaskCleanupType();
+      List<TaskIdWithUsage> shuffleCandidates = getPrioritizedShuffleCandidates(slave);
+
+      for (TaskIdWithUsage task : shuffleCandidates) {
+        if (shufflingRequests.contains(task.getTaskId().getRequestId())) {
+          LOG.debug("Request {} already has a shuffling task, skipping", task.getTaskId().getRequestId());
+          continue;
+        }
+
+        boolean resourceNotOverused = !isOverutilized(slave, cpuUsage, memUsageBytes);
+        boolean tooManyShufflingTasks = isShufflingTooManyTasks(shufflingTasksOnSlave, shufflingTasksOnCluster);
+
+        if (resourceNotOverused || tooManyShufflingTasks) {
+          LOG.debug("Not shuffling any more tasks on slave {} ({} overage : {}%, shuffledOnHost: {}, totalShuffleCleanups: {})",
+              task.getTaskId().getSanitizedHost(),
+              slave.resource.resourceType,
+              slave.resource.overusage * 100,
+              shufflingTasksOnSlave,
+              shufflingTasksOnCluster
+          );
+          break;
+        }
+
+        String message = getShuffleMessage(slave, task, cpuUsage, memUsageBytes);
+        bounce(task, shuffleCleanupType, Optional.of(message));
+
+        cpuUsage -= task.getUsage().getCpusUsed();
+        memUsageBytes -= task.getUsage().getMemoryTotalBytes();
+
+        shufflingTasksOnSlave++;
+        shufflingTasksOnCluster++;
+        shufflingRequests.add(task.getTaskId().getRequestId());
+      }
+    }
+  }
+
+  private List<TaskIdWithUsage> getPrioritizedShuffleCandidates(OverusedSlave slave) {
+    // SingularityUsageHelper ensures that requests flagged as always ineligible for shuffling have been filtered out.
+    List<TaskIdWithUsage> out = slave.tasks;
+
+    if (slave.resource.resourceType == Type.CPU) {
+      out.sort((u1, u2) -> Double.compare(
+          getUtilizationScoreForCPUShuffle(u2),
+          getUtilizationScoreForCPUShuffle(u1)
+      ));
+    } else {
+      out.sort((u1, u2) -> Double.compare(
+          getUtilizationScoreForMemoryShuffle(u1),
+          getUtilizationScoreForMemoryShuffle(u2)
+      ));
+    }
+
+    return out;
+  }
+
+  private double getUtilizationScoreForCPUShuffle(TaskIdWithUsage task) {
+    return task.getUsage().getCpusUsed() / task.getRequestedResources().getCpus();
+  }
+
+  private double getUtilizationScoreForMemoryShuffle(TaskIdWithUsage task) {
+    double memoryUtilization = task.getUsage().getMemoryTotalBytes() / task.getRequestedResources().getMemoryMb();
+    double cpuUtilization = task.getUsage().getCpusUsed() / task.getRequestedResources().getCpus();
+
+    return (memoryUtilization + cpuUtilization) / 2;
```

Comment:
Still thinking more on the metric for this. You were correct to point out the fact that, by shuffling more/smaller tasks we may very well hit the shuffling limits more often. I'm not certain if it will be often enough to be of consequence given the size of our cluster, but there are 2 things I could see us doing to compensate for it:
- Possibly make the shuffle limits a percentage of the overall task count (i.e. so we don't have to worry about it being more restrictive as the cluster grows larger)
- Possibly make a slightly more complex scoring mechanism. I could see a case where there is maybe a threshold for cpu (something simple like a 50-60% cutoff) in memory shuffle. If tasks are _below_ that, they are given a higher priority for high memory usage (inverse of now), if they are above that they are scored as they are now. This could serve to put a few higher memory, but more likely less busy tasks at the top of the queue, resulting in fewer overall shuffles